form/data_transfer_objects.rst

```diff
@@ -0,0 +1,347 @@
+following ``TaskData`` class:
+
+.. code-block:: php
+
+    // src/Form/Data/TaskData.php
+    namespace App\Form\Data;
+
+    use App\Entity\Task;
+    use Symfony\Component\Validator\Constraints as Assert;
+
+    /**
+    * Data transfer object for Task.
+    * Add your constraints as annotations to the properties.
+    */
+    class TaskData
+    {
+        /**
+        * @Assert\NotBlank(message="This value should not be blank.", payload=null)
+        */
+        public $task;
+
+        /**
+        * @Assert\NotBlank(message="This value should not be blank.", payload=null)
+        * @Assert\Type(message="This value should be of type {{ type }}.", type="\DateTime", payload=null)
+        */
+        public $dueDate;
+
+        /**
+        * Create DTO, optionally extracting data from a model.
+        *
+        * @param Task|null $task
+        */
+        public function __construct(? Task $task = null)
+        {
+            if ($task instanceof Task) {
+                $this->extract($task);
+            }
+        }
+
+        /**
+        * Fill entity with data from the DTO.
+        *
+        * @param Task $task
+        */
+        public function fill(Task $task): Task
+        {
+            $task
+                ->setTask($this->task)
+                ->setDueDate($this->dueDate)
+            ;
+
+            return $task;
+        }
+
+        /**
+        * Extract data from entity into the DTO.
+        *
+        * @param Task $task
+        */
+        public function extract(Task $task): self
+        {
+            $this->task = $task->getTask();
+            $this->dueDate = $task->getDueDate();
+
+            return $this;
+        }
+    }
+
+Notice the assert annotations? These were copied from the Task entity.
+The ``extract`` and ``fill`` methods can be used to populate the DTO with data
+from the entity and vice versa.
+
+.. caution::
+
+    During the generation of a DTO, validation annotations are copied from the
+    Entity.
+    You must ensure that changes to the validations are added in both places
+    when the entity is used with forms in other places (like
+    ``SonataAdminBundle`` or ``EasyAdminBundle``).
+    If the entity is not used at all, it is recommended to move all validations
+    into the DTO, removing them from the entity class.
+
+.. index::
+   single: Using the DTO in the Form
+
+Using the DTO in the Form
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use the maker to create a simple CRUD application.
+
+.. code-block:: terminal
+
+    $ php bin/console make:crud Task
+
+This will generate a bunch of templates, a controller and a form.
+First, take a look at the generated ``TaskType`` form.
+
+Notice that it uses the ``Task`` entity by default.
+This means that the form data is injected into the ``Task`` entity directly and validated with the annotations.
+
+Replace this with ``TaskData`` to prevent the aforementioned problems with an invalid entity.
```

Comment:
`aforementioned` => `before mentioned`?

Reply:
I think aforementioned is a nice word and perfectly suitable in this case, as it would have to be `as mentioned before` instead, which is much longer.

Reply:
Hm, I can agree with that.